app/models/roster.rb

```diff
@@ -69,4 +69,12 @@ def user_options
     nas = non_admins.order(:last_name).map { |na| [na.full_name, na.id] }
     { 'Admins' => as, 'Non-Admins' => nas }
   end
+
+  def uncovered_dates_between(start_date, end_date)
+    open_dates = []
+    start_date.to_date.upto(end_date.to_date).each do |date|
+      open_dates << date if Assignment.where(roster: self).between(date, date).empty?
+    end
+    open_dates
```

Comment:
You can do one better than `tap`:

```suggestion
    (start_date.to_date...end_date.to_date).select do |date|
      assignments.between(date, date).empty?
    end
```

Reply:
I'll also offer this up - an optimized implementation:

```suggestion
    (start_date.to_date..end_date.to_date).to_a - assignments.between(start_date, end_date)
                                                             .inject([]) do |dates, assignment|
      dates | (assignment.start_date..assignment.end_date).to_a
    end
```

I noticed that this has a query complexity of `O(n)` with respect to the number of assignments in the date range (each `.empty?` call will produce a SQL query). I couldn't resist tinkering and came up with this single-query version.

It works by collecting all of the covered dates (the `assignment.between(..).inject([])...`), then subtracting them from the full set of dates (i.e. `uncovered_dates = all_dates - covered_dates`).

In reality it doesn't really matter since this is gonna run in the background though, so do with it what you will.